Automattic/vip
# feat(dev-env): Add Photon

## Description

This PR adds support for Photon to the LDE.

It depends on https://github.com/Automattic/vip-container-images/pull/477

## Steps to Test

**NB:** if Automattic/vip-container-images#477 is not merged yet, you will need to build `ghcr.io/automattic/vip-container-images/photon:latest` manually. In this case, you may need to run `vip dev-env start` twice sometimes, as Lando might try to pull `ghcr.io/automattic/vip-container-images/photon:latest` from the remote registry and fail.

1. Check out PR.
2. Run `npm run build`
3. Create an environment with Photon enabled, start it.
4. Upload a test image file.
5. Verify that the file opens in the browser. It should have a `X-Powered-By: PHP/8.1.18` in the response headers.
6. Try adding Photon-specific parameters (like `?w=100`) and make sure they work as expected.

Bonus: try the above with the media redirect domain set. I was unable to set the media redirect domain with `vip dev-env update -r https://my-real-site.com` (could be a bug); if this is the case, you may need to edit `instance_data.json` manually. Restart the environment. The above scenarios should work. If you try to access an image that does not exist, it should be downloaded from `mediaRedirectDomain`.

### End-to-end testing

```sh
vip dev-env create -H -r https://somesite.com < /dev/null
vip dev-env start
wget https://picsum.photos/800 -O ~/.local/share/vip/dev-environment/vip-local/uploads/image.jpg
wget http://vip-local.vipdev.lndo.site/wp-content/uploads/image.jpg -q -O - | identify -
# JPEG 800x800 800x800+0+0
wget 'http://vip-local.vipdev.lndo.site/wp-content/uploads/image.jpg?w=100&filter=blur,grayscale' -q -O - | identify -
# JPEG 100x100 100x100+0+0
```

* `http://vip-local.vipdev.lndo.site/wp-content/uploads/image.jpg`: should show the original image served by nginx; there will be no X-Powered-By header;
* `http://vip-local.vipdev.lndo.site/wp-content/uploads/image.jpg?w=100&filter=blur,grayscale`: will show the image processed by Photon;
* `http://vip-local.vipdev.lndo.site/wp-content/uploads/2019/01/imageforentry1-zci.jpg` will show (redirect to) https://somesite.com/wp-content/uploads/2019/01/imageforentry1-zci.jpg
